_posts/2019-02-25-text-layout-kickoff.markdown

```diff
@@ -0,0 +1,81 @@
+
+The choice of whether to use HarfBuzz or platform shaping is a tradeoff. Using HarfBuzz means more consistency across platforms, and thus an easier testing story. Using platform text means more consistency with other apps on the platform, and less code in the critical path (so shorter compile times and smaller executables).
+
+## Expected users
+
+This work is being funded by Mozilla Research to be used in Servo. Doing text layout for the Web is a very complex problem. The higher level representation of rich text is very tightly bound to the DOM and CSS. All that logic needs to live in Servo, with the lower level providing a clean interface to lay out a span of text with a *single* style and a single [BiDi] direction.
+
+The second major use case is [piet], the 2D graphics abstraction I've been working on. Currently for the Cairo back-end piet uses the "toy text API," which is of course [inadequate][piet#10]. The idea is that skribo handles text layout for Cairo and all future low-level drawing back-ends (including any future back-end using WebRender or direct Vulkan drawing).
+
+Additionally, just about every game engine currently being implemented in Rust uses a simplistic approach to text layout, just looking up each codepoint in the font's cmap and then using its advance width to build a layout. I want to encourage all such engines to migrate to skribo, and want to make that experience smooth. In short, if you're using [rusttype], you should consider using skribo.
+
+## Problems to solve
+
+Aside from being glue abstracting over implementation details, the main actual problem skribo addresses is choosing fonts from a "font collection", which is similar to a "font stack" in CSS. In general, this will contain custom fonts as well as a set of system fallback fonts. This might seem like a relatively simple problem, but there are tons of tricky details – no doubt, I'll write a blog post on just this topic.
+
+Another major problem is making sure locale information affects layout properly. One of the most important to solve is [Han unification], which is the fact that Chinese, Japanese, and Korean share unicode codepoints even though they can be considered different scripts. Often, they should be rendered with different fonts (meaning that Unicode coverage is not the only criterion for selecting fonts). Alternatively, fonts like [Source Han Sans] have variant glyphs for all CJK languages, and use the "locl" OpenType feature to select them. A good layout library handles both, transparently to applications.
+
+Other problems within scope are figuring out logic for "fake bold" and "fake italic" when the font collection doesn't provide true versions and adding [letter-spacing].
```

Comment:
hmm, I would ~versions and~ -> "versions, as well as ..."; and seems ambiguous in this context (is it an additional thing that the font collection doesn't provide?)